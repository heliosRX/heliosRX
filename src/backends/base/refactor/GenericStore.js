class Ref {

  path = ""

  constructor(path) {
    this.path = path;
  }

  push() { // -> move to Db.PushId()
    const PUSHID = 1;
    return PUSHID;
  }

  update( payload ) {
    // !!!!
    return Promise
  }

  remove() {
    return Promise
  }

  transaction(transaction: Function) {
    return Promise

  }

  child() {

  }

  fetchAsObject ({ document, ops, resolve, reject }) {}
  bindAsObject ({ document, ops, resolve, reject }) {}

  fetchAsArray ({ collection, ops, resolve, reject }) {}
  bindAsArray ({ collection, ops, resolve, reject }) {}
}

class Db {
  ref( path = null ) {
    return new Ref(path);
  }
  // generateId()
}

export default {
  childRef(id) {
    return this._db.ref( this.path.replace(/\{id\}/g, id) );
  },

  get parentRef() {
    return this._db.ref( this.path.split('{id}').shift() );
  },

  get rootRef() {
    return this._db.ref();
  },

  // READ MIXIN

  buildQueryRef({
    key = undefined,
    value = undefined,
    limit = undefined,
    startAt = undefined,
    endAt = undefined,
  }) {
  },

  _infer_local_path_from_ref( ref ) {
  },
}
