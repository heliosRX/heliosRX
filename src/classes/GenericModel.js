import { _Vue as Vue } from '../external-deps'
import clonedeep from 'lodash.clonedeep'
// import { _models } from '../external-deps'
import { add_custom_getters } from '../classes/utils'
import moment from '../moment'
import { walkGetPropSave, walkGetObjectSave, walkSetVueProp } from '../registry/utils'

const externalVMStore = new WeakMap(); // Exclude vm from instance, so it can be serialized
const externalModelStore = new WeakMap(); // Stores generic store references

const SERVER_TIMESTAMP_ALIASES = [
  'ServerTimestamp',
  'CurrentTimestamp',
  'InitialTimestamp'
]

export default class GenericModel {

  // ---------------------------------------------------------------------------
  constructor( schema, data, name ) {

    // TODO: Hide internal data from user
    this.$ready       = false;
    /* This promise will resolve once the data is ready */
    this.$promise     = Promise.resolve();
    this.$state       = {};
    this.$dirty       = {};
    this.$invalid     = {};
    this.$id          = null;
    this.$idx         = null;
    this.$noaccess    = null;
    this._store_name  = name;

    Vue.observable( this.$state ); // TODO: Check if we get an error here
    // Vue.observable( this.$ready );
    // this.$state = Vue.oberservable({});
  }

  // ---------------------------------------------------------------------------
  _update_data( data, schema, make_dirty = false ) {
    this.$state = data; // otherwise no reactivity

    if ( make_dirty ) {
      for ( var key in data ) {
        this.$dirty[ key ] = true;
      }
    }

    /*schema.forEach(field => {
      let propName = field.model;
      if ( propName in data ) {
        this.$state[ propName ] = data[ propName ];
      }
    })*/
  }

  // ---------------------------------------------------------------------------
  clone() {
    const copy = Object.assign( Object.create( Object.getPrototypeOf(this)), this);
    copy.$state = clonedeep( copy.$state );
    copy.$dirty = clonedeep( copy.$dirty );
    copy.$invalid = clonedeep( copy.$invalid );
    copy._autogenerate_props( this.$model.modelDefinition.schema.fields, copy.$state, false )
    copy._set_generic_store( this.$model )
    externalVMStore.set( copy, this.$vm );
    return copy;
  }

  // ---------------------------------------------------------------------------
  /* _write_field(key, value) {
    // TODO: We should not alter the state here!!!
    // Create a copy!
    this.$state[ key ] = value;
    this.$dirty[ key ] = true;
  } */

  setValidationBehaviour( value ) {
    // TODO: Use enum
    if ( ['ERROR', 'WARNING', 'ELEMENT_VALIDATION', 'NONE'].includes( value ) ) {
      this._validation_behaviour = value
    }
  }

  _update_field( propName, value, field ) {
    let validation_behaviour = this._validation_behaviour || 'WARNING';

    const prop_set = field.is_nested
                   ? () => walkSetVueProp( this.$state, propName, value )
                   : () => Vue.set( this.$state, propName, value );

    if ( moment.isMoment( value ) ) {
      /* Handle moment object as input */
      if ( !moment.isValidDate( value ) ) {
        console.warn("Got invalid e-moment object", value, "for prop", propName);
      }
      value = value.toRealtimeDB() /* convert to internal timestamp */
      delete this.$invalid[ propName ];
      this.$dirty[ propName ] = true;
      prop_set()
      return
    }

    if ( !field.validator ) {
      /* No validate callback found */

      /* Fixes bug where non existing fields are assigned */
      // Vue.set( this.$state, propName, value );
      prop_set()
      // this.$state[ propName ] = value;

      this.$dirty[ propName ] = true;
      return
    }

    // TODO: bolt type validation

    if ( field.validator( value ) ) {
      // this._write_field( propName, value );
      // this.$state[ propName ] = value;
      delete this.$invalid[ propName ];
      this.$dirty[ propName ] = true;
      prop_set()
    } else {
      // this._write_field( propName, value );
      // this.$state[ propName ] = value;
      this.$invalid[ propName ] = true;
      this.$dirty[ propName ] = true;
      prop_set()

      if ( validation_behaviour === 'ERROR' ) {
        throw new Error(`Validation failed for field '${propName}' with value ${value}.`);
      }
      if ( validation_behaviour === 'WARNING' ) {
        console.warn(`Validation failed for field '${propName}' with value ${value}.`);
      }
      if ( validation_behaviour === 'ELEMENT_VALIDATION' ) {
        // Validation failed, but that's ok - element ui will take care
      }
    }
  }

  // ---------------------------------------------------------------------------
  _on_remove() {
    this.$deleted = true; // TODO
    // This is called when the item is going to be removed
    //  - maybe allow subscribers and call them here
  }

  // ---------------------------------------------------------------------------
  $isValid() {
    return Object.keys( this.$invalid ).length === 0;
  }

  // ---------------------------------------------------------------------------
  get $key() {
    return 'KEY-' + this._store_name + '-' + this.$id;
  }

  // ---------------------------------------------------------------------------
  get $vm() {
    return externalVMStore.get( this )
  }

  // ---------------------------------------------------------------------------
  get $model() {
    return externalModelStore.get( this )
  }

  // ---------------------------------------------------------------------------
  get $exists() {
    return this.$state[ '.exists' ] !== false;
  }

  // ---------------------------------------------------------------------------
  _autogenerate_props( schema, data, is_dirty = false ) { // TODO: move to util

    if ( !Array.isArray( schema ) ) {
      schema = Object.keys(schema).map(key => {
        return { model: key, ...schema[key] }
      });
    }

    schema.forEach(field => {
      let propName = field.model;
      // console.log("[_autogenerate_props]", field, propName, data)

      if ( data && propName in data ) {
        this.$state[ propName ] = data[ propName ];
      }
      /* else {
        this.$state[ propName ] = field.placeholder;
      } */

      if ( is_dirty ) {
        this.$dirty[ propName ] = true;
      }

      if ( Object.prototype.hasOwnProperty.call(this, propName ) ) {
        console.warn(`Name conflict: property "${propName}" has same name as global action/global getter "${propName}" in ${this._store_name}`);
        return
      }

      let prop_getter = () => {};
      let prop_path = '';

      if ( propName.includes('/') || propName.includes('.') ) {
        prop_path = propName.replace(/\//g, '.')
        let prop_keys = prop_path.split('.')
        let prop_last_key = prop_keys.pop()
        prop_keys = prop_keys.join('.')

        prop_getter = () => {
          // return walkGetPropSave( this.$state, prop_path )
          let parent = walkGetObjectSave( this.$state, prop_keys )
          return parent[ prop_last_key ]
        }
        field.is_nested = true
      } else {
        prop_path = propName
        prop_getter = () => {
          return this.$state[ propName ];
        }
        field.is_nested = false
      }

      if ( field.type ) {

        const validate_timestamp = (value, moment_conversion_func, min_date, max_date) => {
          if ( value === 0 || value === null ) {
            return value // TODO: Allow null instead of 0
          } else if ( moment.isMoment( value ) ) {
            console.error("Do not assign moment objects directly.");
          } else if ( isFinite( value ) && value > min_date && value < max_date ) {
            return moment_conversion_func( value );
          } else if ( value === undefined ) {
            return moment_conversion_func(); // TODO: This is potentially dangerous, since it returns the current time
          } else {
            console.warn("Schema defined", propName, "as Timestamp, but got invalid data:", value);
            return value
          }
        }

        let prop_getter_original = prop_getter;

        if ( field.type === 'Timestamp' ) {

          prop_getter = () => {
            const min_date =   200000000 // '1976-05-03'
            const max_date = 30000000000 // '2065-01-24'
            return validate_timestamp(
              prop_getter_original(),
              moment.fromRealtimeDB,
              min_date,
              max_date)
          }
        } else if ( SERVER_TIMESTAMP_ALIASES.includes( field.type ) ) {
          prop_getter = () => {
            const min_date =   200000000 * 1000 // '1976-05-03'
            const max_date = 30000000000 * 1000 // '2065-01-24'
            return validate_timestamp(
              prop_getter_original(),
              moment.fromRealtimeServerTime,
              min_date,
              max_date)
          }
        } else if ( field.type.includes( 'Timestamp' ) ) {
          console.warn(
            "Found validation type that contains 'Timestamp' but is not recognized:",
            field.type
          );
        }

        // TODO: Also validate other types
      }

      Object.defineProperty(this, prop_path, {
        set: value => this._update_field( prop_path, value, field ),
        get: prop_getter
      });
    });
  }

  // ---------------------------------------------------------------------------
  _decorate_actions( modelActions, context ) { // TODO: move to util
    if ( !modelActions ) {
      return
    }
    context.model = this; // HACK
    for ( let key in modelActions ) {
      let action = modelActions[ key ];
      if ( Object.prototype.hasOwnProperty.call( this, key ) ) {
        console.warn(`Name conflict: action "${key}" has same name as property/global action/global getter "${key}" in ${this._store_name}`);
        continue
      }
      Object.defineProperty( this, key, { value: () => action(context) } ) // TODO: bind this?
    }
  }

  // ---------------------------------------------------------------------------
  _decorate_getters( modelGetters, context ) { // TODO: move to util
    if ( !modelGetters ) {
      return
    }

    for ( let key in modelGetters ) {
      if ( Object.prototype.hasOwnProperty.call( this, key ) ) {
        console.warn(`Name conflict: getter "${key}" has same name as property/custom action/global action/global getter "${key}" in ${this._store_name}`);
        delete modelGetters[ key ] // ?
        continue
      }
    }

    /* Embed getter in vue instance */
    context.model = this; // HACK
    let vm = add_custom_getters( context, this, modelGetters );
    externalVMStore.set( this, vm );
  }

  // ---------------------------------------------------------------------------
  _set_generic_store( store ) {
    externalModelStore.set( this, store );
  }

  // ---------------------------------------------------------------------------
  write() {
    // TODO: Nested data

    console.log("Writing $dirty fields", JSON.stringify(this.$dirty));

    let payload = {}
    for ( let prop in this.$dirty ) {
      let value
      let prop_path
      if ( prop.includes('.') ) {
        prop_path = prop.replace(/\./g, '/')
        value = walkGetPropSave( this.$state, prop );
      } else {
        prop_path = prop
        value = this.$state[ prop ];
      }
      if ( typeof value === 'undefined' ) {
        console.warn("Trying to write undefined for prop", prop);
        continue
      }
      payload[ prop_path ] = value;
    }

    let model = this.$model;
    if ( !model ) {
      throw new Error('Model reference not set with _set_generic_store');
    }

    let is_update = true;
    let temp_id = this.$id;
    if ( !temp_id ) {
      temp_id = model._get_uid();
      is_update = false;
    }

    /* Check if all required fields are present */
    if ( model.modelDefinition && model.modelDefinition.schema ) {
      model._validate_schema( payload, is_update ) // DOPPELT ?
    } else {
      console.warn("No schema found to validate input", );
    }

    return model.update( temp_id, payload ).then(() => {
      this.$id = temp_id;
      this.$dirty = {}
      return temp_id;
    });

    /* 'store.add' would call 'create' - not what we want here */
    // TODO: Alternative { directWrite: true }
  }

  /* ------------------------------------------------------------------------ */
  _get_model_for_write() {
    if ( !this.$id ) {
      throw new Error('Write operations are not allowed for new models.');
    }

    let model = this.$model;
    if ( !model ) {
      throw new Error('Model reference not set with _set_generic_store');
    }
    return model
  }

  /* ------------------------------------------------------------------------ */
  update( payload ) {
    let model = this._get_model_for_write();
    return model.update( this.$id, payload ).then(() => {
      for ( var propName in payload ) {
        delete this.$dirty[ propName ]
      }
      return this.$id;
    });
  }

  /* ------------------------------------------------------------------------ */
  remove(soft_delete = true) {
    let model = this._get_model_for_write();
    return model.remove( this.$id, soft_delete ).then(() => {
      if ( soft_delete ) {
        delete this.$dirty[ 'deleted' ] // ?
      }
      // TODO Mark model as deleted?
    })
  }

  /* ------------------------------------------------------------------------ */
  restore() {
    let model = this._get_model_for_write();
    return model.restore( this.$id ).then(() => {
      delete this.$dirty[ 'deleted' ] // ?
    })
  }

  // ---------------------------------------------------------------------------
  // hotUpdate() {}

  // -----------------------------------------------------------------------------
  reset() {
    // console.log("[GENS] Reset Model -> NOT IMPLEMENTED")
    // ....
  }

  // ===========================================================================
  /*
  customActionExample() {
    console.log("custom action 1");
  }
  */
}
