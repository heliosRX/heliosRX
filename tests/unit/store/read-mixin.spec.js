import GenericStore from '@/store'
import GenericModel from '@/classes/GenericModel'
import { MockRef, MockBackend } from './mock-backend'
import { matcherHint, printReceived } from 'jest-matcher-utils';
import registryModule from '@/registry/module'
import setupExternalDeps from '@/external-deps'
import { setup as storeSetup } from '@/store/GenericStore.js'
import {
  rtdbFetchAsObject,
  rtdbFetchAsArray,
  rtdbBindAsObject,
  rtdbBindAsArray,
} from '@/backends/firebase-rtdb/rtdb.js'

import Vue from 'vue'
// import Vue from 'vue/dist/vue.esm.js'
import Vuex from 'vuex'
Vue.config.productionTip = false;
Vue.config.devtools = false;

jest.mock('./mock-backend');
jest.mock('@/backends/firebase-rtdb/rtdb.js');

/* eslint-disable quote-props */
const modelDefinition = {
  schema: {
    create({ strInput }) {
      return {
        strField: strInput || "Empty",
        numField: 8,
      }
    },
    fields: {
      timestampField: { type: 'Timestamp' },
      strField: { type: 'String' },
      boolField: { type: 'Boolean' },
      numField: { type: 'Number' },
      mapField: { type: 'Map<String, String>' },
      arrField: { type: 'String[]' },
      'nested.field': { type: 'String' },
      'nested.sub.field': { type: 'String' },
      "/dynField[0-9]+/": { type: "String" },
      // Special fields
      deleted: { type: 'Boolean' },
      sortidx: { type: 'Number' },
    },
  }
};

const store = new GenericStore(
  "/example/sub/*",
  modelDefinition
  // { enableTypeValidation: true }
);

const storeNested = new GenericStore(
  "/example/{testId}/sub/*",
  modelDefinition
);

const storeSuffixed = new GenericStore(
  "/example/sub/*/meta",
  modelDefinition
);

const mockBackend = new MockBackend();
const mockContext = {}
let registry = null;

beforeAll(() => {
  GenericStore.setDefaultDB( mockBackend );
  GenericStore._set_caller( mockContext )

  setupExternalDeps({
    Vue,
    // models: options.models,
    // db:     options.db,
  });

  storeSetup({
    Vue,
    // firebase: options.firebase || options.db.app.firebase_ // HACK: Figure out from rtdb
  })

  // Setup registry
  Vue.use( Vuex );
  registry = new Vuex.Store( registryModule( 'heliosRX' ) )
  setupExternalDeps({ Vuex, registry: registry });
  registry.commit('INIT_REGISTRY');
});

beforeEach(() => {
  // console.log("beforeEach - 1")
  MockBackend.mockClear();
  store.resetGlobalInstanceCache();
  registry.commit('RESET_REGISTRY');
  registry.commit('INIT_REGISTRY');
});

/*
rtdbFetchAsObject.mockImplementation(() => {
  console.log("rtdbFetchAsObject")
});
rtdbFetchAsArray.mockImplementation(() => {
  console.log("rtdbFetchAsArray")
});
rtdbBindAsObject.mockImplementation(() => {
  console.log("rtdbBindAsObject")
});
// rtdbFetchAsObject.mockClear();
// rtdbFetchAsArray.mockClear();
// rtdbBindAsObject.mockClear();
*/

describe('subscribeList / subscribeQuery', () => {
  // subscribeList( <idList: array<string>> <, option:object>) ⇒ GenericList
  // subscribeQuery( <query:QueryDefinition> <, option:object>) ⇒ GenericList

  let mockRef = null;
  let mockRefList = [];
  const mockHelpers = {
    setList: null,
    addItem: null,
    removeItem: null,
    changeItem: null,
  }

  beforeAll(() => {
    mockBackend.ref.mockImplementation((path) => {
      const ref = new MockRef(path)
      ref.constructor.mockImplementation((path) => {
        // console.log("PATH", path)
      })
      ref.update.mockResolvedValue(null)
      ref.path = path;
      mockRef = ref;
      mockRefList.push(ref)
      return ref
    });

    rtdbBindAsArray.mockImplementation(({ collection, ops, resolve, reject }) => {
      const { target } = ops.init()

      mockHelpers.addItem = (key, data) => {
        ops.add( target, key, data )
      }
      mockHelpers.removeItem = (key) => {
        ops.remove( target, key )
      }
      mockHelpers.changeItem = (key, data) => {
        ops.set( target, key, data )
      }
      mockHelpers.setList = (data, exists = true) => {
        ops.once( target, data, exists )
        resolve();
      }

      return () => {
        console.log("UNSUBSCRIBE")
      }
    });
  });

  beforeEach(() => {
    // console.log("beforeEach - 2")
    mockBackend.ref.mockClear();
    if ( mockRef ) {
      mockRef.constructor.mockClear();
      mockRef.push.mockClear();
      mockRef.update.mockClear();
      rtdbBindAsArray.mockClear();
    }
  });

  test('subscribeList 1', async () => {

    const list = store.subscribeList();

    const data = {
      'A': { numField: 9 },
      'B': { numField: 11 },
    };

    setTimeout(() => {
      mockHelpers.setList( data );
    }, 10)

    setTimeout(() => {
      Object.keys(data).forEach(key => {
        mockHelpers.addItem( key, data[key] );
      })
    }, 20)

    await new Promise(res => setTimeout(res, 30))

    await list.$promise

    /*
    console.log(list.items)
    console.log(list.items['A'].$state)
    console.log(list.$promise)
    console.log(registry.state.initialized)
    console.log(registry.state.ready)
    console.log(registry.state.res)
    console.log(registry.state.res.example.sub)
    console.log(JSON.stringify(registry.state, null, 2))
    */

    expect( registry.state.sync ).toHaveProperty('/example/sub/{id}')
    expect( registry.state.sync['/example/sub/{id}'].query ).toBe(null)
    expect( registry.state.sync['/example/sub/{id}'].status ).toBe('Ready')

    expect( list.$idList.length ).toBe(2)
    expect( list.items['A'].$state ).toEqual({ numField: 9 })
    expect( list.items['B'].$state ).toEqual({ numField: 11 })
    expect( registry.state.res.example.sub.A ).toEqual({ numField: 9 })

    mockHelpers.addItem("C", { strField: "foo" });

    await new Promise(res => setTimeout(res, 10))

    expect( list.items['C'].$state ).toEqual({ strField: "foo" })
    expect( registry.state.res.example.sub.C ).toEqual({ strField: "foo" })

    let list_from_cache = store.subscribeList();
    expect( list_from_cache.$idList ).toHaveLength(3)
    expect( list_from_cache ).toBe( list ) // Check for referential identity

    mockHelpers.removeItem('A');
    expect( list.items ).not.toHaveProperty('A')
    // expect( list.$numReady ).toBe(2) // TODO

    // mockHelpers.addItem(key, data);
    mockHelpers.changeItem('B', { numField: 250 });
    await new Promise(res => setTimeout(res, 30))

    // TODO !!
    /*
    console.log(list.items['B'].$state)
    console.log(JSON.stringify(registry.state, null, 2))
    expect( list.items['B'].$state ).toEqual({ numField: 25 })
    */
  });

  test.todo('subscribeList with already synced parent')

  test('subscribeQuery 1', async () => {

    let mockRefs = [];

    mockBackend.ref.mockImplementation((path) => {
      const ref = new MockRef(path)
      ref.orderByValue.mockReturnValue(ref)
      ref.orderByKey.mockReturnValue(ref)
      ref.orderByChild.mockReturnValue(ref)
      ref.equalTo.mockReturnValue(ref)
      ref.startAt.mockReturnValue(ref)
      ref.endAt.mockReturnValue(ref)
      ref.limitToFirst.mockReturnValue(ref)
      ref.limitToLast.mockReturnValue(ref)
      // ref.update.mockResolvedValue(null)
      ref.path = path;
      mockRefs.push(ref);
      return ref
    });

    // SELECT * WHERE numField < 20 LIMIT 3
    const list = store.subscribeQuery({ key: 'numField', endAt: 20, limit: 3 });

    let mockRefQuery = mockRefs[0];

    await new Promise(res => setTimeout(res, 30))

    expect( mockRefQuery.endAt ).toHaveBeenCalledWith(20)
    expect( mockRefQuery.orderByChild ).toHaveBeenCalledWith('numField')
    expect( mockRefQuery.limitToFirst ).toHaveBeenCalledWith(3)

    const data = {
      'A': { numField: 9 },
      'B': { numField: 11 },
      'C': { numField: 15 },
      'D': { numField: 25 },
      'E': { numField: 95 },
    };

    setTimeout(() => {
      mockHelpers.setList( data );
    }, 10)

    setTimeout(() => {
      mockHelpers.addItem( 'A', data['A'] );
      mockHelpers.addItem( 'B', data['B'] );
      mockHelpers.addItem( 'C', data['C'] );
    }, 20);

    await new Promise(res => setTimeout(res, 30))
    await list.$promise;

    expect( list.$idList.length ).toBe(3)
    expect( list.items['A'].$state ).toEqual(data['A'])
    expect( list.items['B'].$state ).toEqual(data['B'])
    expect( list.items['C'].$state ).toEqual(data['C'])

    list.itemsAsArray().map( item => {
      expect( item.numField ).toBeLessThan(20)
    })
  })
});

describe('subscribeNode', () => {
  // subscribeNode( <id: string> <, option:object>) ⇒ GenericModel

  let mockRef = null;
  const mockHelpers = {
    setItem: null,
  }

  beforeAll(() => {
    // console.log("beforeAll - 2")
    mockBackend.ref.mockImplementation((path) => {
      console.log('* creating ref with path <' + path + '> *')
      const ref = new MockRef(path)
      ref.constructor.mockImplementation((path) => {
        console.log('* creating NEW ref with path <' + path + '> *')
      })
      ref.path = path;
      mockRef = ref;
      return ref
    });

    rtdbBindAsObject.mockImplementation(({ collection, ops, resolve, reject }) => {

      const { target } = ops.init()

      mockHelpers.setItem = (data, exists = true) => {
        ops.set( target, data ) // TODO: Also pass { .exists } here?
        resolve(data);
      }

      return () => {
        console.log("UNSUBSCRIBE")
      }
    });
  });

  beforeEach(() => {
    console.log("beforeEach - 2")
    mockBackend.ref.mockClear();
    if ( mockRef ) {
      mockRef.constructor.mockClear();
      mockRef.push.mockClear();
      rtdbBindAsObject.mockClear();
    }
  });

  test('subscribeNode 1', async () => {

    const node = store.subscribeNode('id1');

    const data = { numField: 9, strField: 'foo', boolField: true };

    setTimeout(() => {
      mockHelpers.setItem( data );
    }, 10)

    await node.$promise

    expect(mockRef.path).toBe('/example/sub/id1')

    expect(node.$ready).toBe(true)
    expect(node.$dirty).toEqual({})
    expect(node.$noaccess).toBe(false)

    expect( store.getData( 'id1' ) ).toEqual( data )
    expect( node.$state ).toEqual( data )

    expect( registry.state.res.example.sub['id1'] ).toEqual(data);
    expect( registry.state.sync ).toEqual({
      "/example/sub/id1": {
        "id": "id1",
        "status": "Ready"
      }
    })
    expect( registry.state ).toMatchSnapshot()

    expect( store.getAllSyncedPaths() ).toEqual({ '/example/sub/id1': 'Ready' })

    // console.log(JSON.stringify(registry.state, null, 2))
  });
});

describe.skip('fetchNode', () => {
  // fetchNode( <id:string>, <option:object> ) ⇒ GenericModel
});

describe.skip('fetchQuery', () => {
  // fetchQuery( <query:QueryDefinition> <, option:object> ) ⇒ GenericList
});

describe.skip('getList', () => {
  // getList( <idList:array<string>> <, option:object>) ⇒ GenericList
});

describe.skip('getNode', () => {
  // getNode( <id:string> <, option:object>) ⇒ GenericModel
});

describe.skip('getData', () => {
  // getData( <id:string> <, option:object>) ⇒ Object<string:any>
});

/*
describe.skip('exists', () => {
  // exists( <id:string> ) ⇒ boolean
});

describe.skip('getRegistryState', () => {
  // getRegistryState() => object
});

describe.skip('getAllSyncedPaths', () => {
  // getAllSyncedPaths() => object<string:string>
});
*/
