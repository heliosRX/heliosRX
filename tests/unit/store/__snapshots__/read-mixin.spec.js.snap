// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`subscribeNode subscribeNode 1 1`] = `
Object {
  "initialized": true,
  "ready": Object {},
  "res": Object {
    "example": Object {
      "sub": Object {
        "id1": Object {
          "boolField": true,
          "numField": 9,
          "strField": "foo",
        },
      },
    },
  },
  "sync": Object {
    "/example/sub/id1": Object {
      "id": "id1",
      "status": "Ready",
    },
  },
}
`;
